core/src/main/java/io/questdb/griffin/engine/functions/str/SubStringFunctionFactory.java

```diff
@@ -0,0 +1,162 @@
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.str;
+
+import io.questdb.cairo.CairoConfiguration;
+import io.questdb.cairo.TableUtils;
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.functions.StrFunction;
+import io.questdb.griffin.engine.functions.TernaryFunction;
+import io.questdb.griffin.engine.functions.constants.StrConstant;
+import io.questdb.std.IntList;
+import io.questdb.std.Numbers;
+import io.questdb.std.ObjList;
+import io.questdb.std.str.StringSink;
+import org.jetbrains.annotations.Nullable;
+
+public class SubStringFunctionFactory implements FunctionFactory {
+    @Override
+    public String getSignature() {
+        return "substring(SII)";
+    }
+
+    @Override
+    public Function newInstance(final int position, final ObjList<Function> args, IntList argPositions, final CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) throws SqlException {
+        final Function strFunc = args.getQuick(0);
+        final Function startFunc = args.getQuick(1);
+        final Function lenFunc = args.getQuick(2);
+        if (strFunc.isNullConstant()
+                || startFunc.isConstant() && startFunc.getInt(null) == Numbers.INT_NaN
+                || lenFunc.isConstant() && lenFunc.getInt(null) == Numbers.INT_NaN) {
+            return StrConstant.NULL;
+        }
+        if (lenFunc.isConstant()) {
+            int len = lenFunc.getInt(null);
+            if (len < 0) {
+                throw SqlException.$(position, "negative substring length is not allowed");
+            } else if (len == 0) {
+                return StrConstant.EMPTY;
+            }
+        }
+        return new SubStringFunc(strFunc, startFunc, lenFunc);
+    }
+
+    private static class SubStringFunc extends StrFunction implements TernaryFunction {
+
+        private final Function strFunc;
+        private final Function startFunc;
+        private final Function lenFunc;
+
+        private final StringSink sinkA = new StringSink();
+        private final StringSink sinkB = new StringSink();
+
+        private final boolean isSimplifiable;
+
+        public SubStringFunc(Function strFunc, Function startFunc, Function lenFunc) {
+            this.strFunc = strFunc;
+            this.startFunc = startFunc;
+            this.lenFunc = lenFunc;
+
```

Comment:
This condition could be checked in newInstance and a simplified function could be returned .